core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -235,6 +270,57 @@ class PartitionStateMachine(controller: KafkaController) extends Logging {
     }
   }
 
+
+  /**
+    * Asynchronous version of handleStateChange only for the case when targetState == OnlinePartition.
+    * For any other transitions the syncronous version should be used
+    */
+  private def handleStateChangeAsync(topic: String, partition: Int, targetState: PartitionState,
+                                     leaderSelector: PartitionLeaderSelector,
+                                     topCallback: (scala.Any) => Unit, topCtx: scala.Any) {
```

Comment:
@ijuma : in async functions, the context is given by the caller and is supposed to be opaque to the callee. The callee simply passes the context back to the caller during the callback without inspecting it. I'll see if there are instances where we can avoid scala.Any, but not in callbacks.